### PR TITLE
Support TypeChain in deployProxy function

### PR DESCRIPTION
**Pull request by frimoldi**
As discussed in https://github.com/OpenZeppelin/openzeppelin-upgrades/pull/325, this change allows `deployProxy` to returned a typed `Contract` instance, so the user can use things like Typechain without loosing contract's types definition.

Usage example in a hardhat script 👇🏽
```
import { ethers, upgrades } from "hardhat";
import { Greeter__factory } from "../typechain";

async function main() {
  const signer = (await ethers.getSigners())[0];
  const factory = new Greeter__factory(signer);
  const greeter = await upgrades.deployProxy<Greeter__factory>(factory, {
    constructorArgs: ["Fran"],
  });

  await greeter.deployed();

  // this is now type safe
  await greeter.greet();
}
```

### CODE REVIEW

**Comment by frimoldi** [comment]
@frangio please let me know your thoughts on this.

File: packages/plugin-hardhat/src/deploy-proxy.ts [hunk 1]
```diff
@@ -20,17 +20,19 @@ import {
   getInitializerData,
 } from './utils';
 
+type InstanceOf<F extends ContractFactory> = ReturnType<F['attach']>;
```
Comment (frangio) [inline packages/plugin-hardhat/src/deploy-proxy.ts:23]: Can you call this `InstanceOfFactory` or something with Factory?

**Review by frangio — CHANGES_REQUESTED** [review]
Thank you @frimoldi! This looks solid. I'd love to have a test for this but our tests in this package are written in JavaScript so I don't think we can set one up.

File: packages/plugin-hardhat/src/deploy-proxy.ts (hunk 1 above)
Comment (frimoldi) [reply packages/plugin-hardhat/src/deploy-proxy.ts:23]: yeah, good call.

**Review by frangio — COMMENTED** [review]
Sorry for the delay. I realized we also have to cover the other functions like `upgradeProxy`, `deployBeacon`, etc. Do you want to take a look at those?

**Comment by frimoldi** [comment]
@frangio sure, I'll take a look as soon as I can.

**Comment by frangio** [comment]
@frimoldi Let us know if you can carry on with this. :slightly_smiling_face:

**Comment by frimoldi** [comment]
@frangio Yes. Sorry for the delay, I'll find some time this week.

**Comment by 0xakihiko** [comment]
Would love this

**Comment by 0xWhiteleaf** [comment]
Any update on this?

**Comment by bulbazavr1024** [comment]
@frimoldi Hey! How it's going ?)

**Comment by frankbonnet** [comment]
What's the status of this?

**Comment by ericglau** [comment]
Completed in #1099 and available in [@openzeppelin/hardhat-upgrades@3.8.0](https://github.com/OpenZeppelin/openzeppelin-upgrades/releases/tag/%40openzeppelin%2Fhardhat-upgrades%403.8.0)

**Pull request closed without merging**